docs: rewrite introduction to dependency injection (#44466)
angular/angular: aio/content/guide/inject-object.md

# Injecting an object

To inject an object, configure the injector with the `useValue` option.
The following provider object uses the `useValue` key to associate the variable with the `Logger` token.

<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-7"></code-example>

In this example, `SilentLogger` is an object that fulfills the logger role.

<code-example path="dependency-injection/src/app/providers.component.ts" region="silent-logger"></code-example>

{@a non-class-dependencies}

## Injecting a configuration object

> A common use case for object literals is a configuration object.

Reviewer: ```suggestion
A common-use case for object literals is a configuration object.
```

The following configuration object includes the title of the application and the address of a web API endpoint.

<code-example path="dependency-injection/src/app/app.config.ts" region="config" header="src/app/app.config.ts (excerpt)"></code-example>

To provide and inject the configuration object, specify the object in the `@NgModule()` `providers` array.

<code-example path="dependency-injection/src/app/app.module.ts" region="providers" header="src/app/app.module.ts (providers)"></code-example>

{@a injectiontoken}

## Using an `InjectionToken` object

> Define and use an `InjectionToken` object for choosing a provider token for non-class dependencies. The following example defines a token, `APP_CONFIG` of the type `InjectionToken`.

Reviewer: ```suggestion
Define and use an `InjectionToken` object for choosing a provider token for non-class dependencies. The following example defines a token, `APP_CONFIG`, of the type `InjectionToken`.
```

<code-example path="dependency-injection/src/app/app.config.ts" region="token" header="src/app/app.config.ts"></code-example>

The optional type parameter, `<AppConfig>`, and the token description, `app.config`, specify the token's purpose.

> Next, register the dependency provider in the component using the `InjectionToken` object of `APP_CONFIG`.

Reviewer: ```suggestion
Next, register the dependency provider in the component by using the `InjectionToken` object of `APP_CONFIG`.
```

<code-example path="dependency-injection/src/app/providers.component.ts" header="src/app/providers.component.ts" region="providers-9"></code-example>

Now, inject the configuration object into the constructor with `@Inject()` parameter decorator.

<code-example path="dependency-injection/src/app/app.component.2.ts" region="ctor" header="src/app/app.component.ts"></code-example>

{@a di-and-interfaces}

### Interfaces and dependency injection

Though the TypeScript `AppConfig` interface supports typing within the class, the `AppConfig` interface plays no role in dependency injection.

> In TypeScript, an interface is a design-time artifact, and doesn't have a runtime representation, or token, that the DI framework can use.

Reviewer: ```suggestion
In TypeScript, an interface is a design-time artifact and doesn't have a runtime representation, or token, that the DI framework can use.
```

When the transpiler changes TypeScript to JavaScript, the interface disappears because JavaScript doesn't have interfaces.

Because there is no interface for Angular to find at runtime, the interface cannot be a token, nor can you inject it.

<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-9-interface"></code-example>

<code-example path="dependency-injection/src/app/providers.component.ts" region="provider-9-ctor-interface"></code-example>

{@a factory-provider}
{@a factory-providers}

## Using factory providers

To create a changeable, dependent value based on information unavailable before run time, use a factory provider.

In the following example, only authorized users should see secret heroes in the `HeroService`.

> Authorization can change during the course of a single application session, as when a different user logs in .

Reviewer: ```suggestion
Authorization can change during the course of a single application session, such as when a different user logs in .
```

> To keep security-sensitive information in `UserService` and out of `HeroService`, give the `HeroService` constructor a boolean flag to control display of secret heroes.

Reviewer: ```suggestion
To keep security-sensitive information in `UserService` and out of `HeroService`, give the `HeroService` constructor a Boolean flag to control display of secret heroes.
```

> # Configuring a dependency using DI tokens

Reviewer 2: The title of this section makes sense, as it talks about an abstract dependency, but the link from the main DI page mentions "Injecting an object" and this file is actually named `inject-object.md`. I think we should change those "inject object" references to "Inject a dependency using DI token" or in case of a file name (and corresponding URL): "inject-using-di-tokens.md".

> # Configuring a dependency using DI tokens

Reviewer 2: Note: I think the file should be renamed from "inject-object.md" to something else, since the content talks about other types of content aside from objects.

Reviewer 2: Btw, I don't think we had that page at all in our Google Doc.

Author: You're right. This is an obsolete file which I kept in the repo but removed from the navigation, in case we need to reuse some content from this file. I just removed it from the repo.

> You define as an Angular injection token to identify a category of dependencies. Injecting an object Dis an example of configuring a dependency using DI tokens. 

Reviewer 2: ```suggestion
You define as an Angular injection token to identify a category of dependencies. You can use an injection token to inject a value of any type (object, string, boolean, etc). This guide contains examples on how to use an injection token to provide and inject an object, but the same technique can be used for other types as well.
```
